blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+Lastly, we don't want or expect our users to be expert bug reporters. They are
+often working scientists who are good programmers, but may not have any
+background in software engineering. These are some of our most valuable users,
+and we want to make sure their bugs are addressed.
+
+For users who did encounter such particularly difficult problems, we have for
+a long time had one particular answer: If you can reproduce it on a linux
+machine and get us a trace from the `rr` tool <https://rr-project.org/>, we can
+probably get it fixed for you very rapidly. For the uninitiated, `rr` is a Linux
+debugging tool originally written by Robert O'Callahan at Mozilla. It is a tool
+known as a "time traveling debugger" or "reverse execution engine". Essentially,
+`rr` splits reproducing bugs into a "record" and a "replay" phase. The record
+phase is performed by the bug reporter. During this phase, `rr` creates a perfect
+record of the execution, including the *bitwise exact* memory and register state
+after every instruction. This trace can then be analyzed during the replay
+phase (which can be performed by a different developer on a different machine).
+Capabilities like these have long been imagined in academia, but it is extremely
+hard to achieve in a way that does not introduce large overheads or distort
+regular execution. `rr` is the first such tool that (in our experience) is
+performant enough to be used in the regular course of development. It is worth
+discussing how this is achieved (and what limitations the approach has), but
+first let's take the capabilities for granted and take a look at the workflow
+it enables.
+
+In Julia 1.5, there is now a new command line flag `--bug-report=rr` that will
+automatically create and upload an `rr` recording. An example usage is shown
+in the animation at the start of this post (which just creates a deliberate
+crash by unsafely dereferencing a bad pointer). However to summarize:
+
+1. The bug reporter passes `--bug-report=rr` to her julia instance and
+   reproduces whatever bug they are attempting to reproduce.
+
+2. Once julia exits or crashes, the bug reporter is prompted to authorize an
+   upload by clicking on a link (GitHub-based authentication is used for abuse
+   protection). She then gets back a link to include when filing a bug report
+   against julia or some other package.
+
+3. Any developer can use the link to obtain the recording and analyze it on
+   their own machine.
+
+In addition to this manual mechanism, we are also switching our linux CI systems
+to automatically create `rr` traces of any execution. That way, if a CI run fails,
+we are guaranteed to be able to debug it.
+
+If a bug report includes a link to an `rr` trace, no further reproduction
+instructions are required (if the bug is something non-obvious like
+unexpected behavior, some comments on what the expected behavior was may
+still be helpful), since the `rr` trace is guaranteed to perfectly capture
```

Comment:
I think I'd still want the bug report to still contain the usual suspects (what was the expected result and unexpected observation). Could be helpful in case the `rr` trace is unnecessary (clear from the final backtrace, for example) or breaks or just needs triage, and so on.

Reply:
Well, part of the point is that it should be possible to automatically extract all the useful information from the trace and post it automatically - versioninfo, manifests, etc, even if you don't want to do a full replay. https://github.com/JuliaLang/julia/issues/35601 was a MVP of that.